createUIAlertController title gets truncated if too long

youennf opened the pull request:
#### 7753b1fdea4b4605b4851dec61966b8b7247302a
<pre>
createUIAlertController title gets truncated if too long
<a href="https://bugs.webkit.org/show_bug.cgi?id=264568">https://bugs.webkit.org/show_bug.cgi?id=264568</a>
<a href="https://rdar.apple.com/116345923">rdar://116345923</a>

Reviewed by Jean-Yves Avenard.

UIKit is truncating titles when they are long, which is probablematic for prompts.
We reuse the same workaround as Safari does for its own prompts.

* Source/WebKit/Platform/spi/ios/UIKitSPI.h:
* Source/WebKit/UIProcess/ios/UIKitUtilities.mm:
(WebKit::createUIAlertController):

Canonical link: <a href="https://commits.webkit.org/270760@main">https://commits.webkit.org/270760@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0b0b3d647a73d191b5aa132a54279b808171c670

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/26182 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/4793 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/27457 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/28279 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/23971 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/6576 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/2215 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/23997 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/26439 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/3650 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/22528 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/28854 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/3252 "Passed tests") | | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/29544 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/23872 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/23883 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/27435 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/3292 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/1502 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/4700 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6327 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/3759 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/3614 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebKit/UIProcess/ios/UIKitUtilities.mm:
```diff
@@ -202,8 +202,15 @@ - (UIViewController *)_wk_viewControllerForFullScreenPresentation
-    auto *alert = [UIAlertController alertControllerWithTitle:title message:message preferredStyle:UIAlertControllerStyleAlert];
-    [alert _setTitleMaximumLineCount:0]; // No limit, we need to make sure the title doesn't get truncated.
+    // FIXME: revisit this once UIKit bug is resolved (see also rdar://101140177).
+    NSMutableAttributedString *alertTitle = [[NSMutableAttributedString alloc] initWithString:title];
```
```suggestion
    auto alertTitle = adoptNS([[NSMutableAttributedString alloc] initWithString:title]);
```

youennf replied:
Thanks for the review.
Fixed

pxlcoder commented on Source/WebKit/UIProcess/ios/UIKitUtilities.mm:
```diff
@@ -202,8 +202,15 @@ - (UIViewController *)_wk_viewControllerForFullScreenPresentation
+    alert.attributedTitle = alertTitle;
+    alert.title = title;
+    alert.message = message;
+    alert.preferredStyle = UIAlertControllerStyleAlert;
```
```suggestion
    auto alert = adoptNS([[UIAlertController alloc] init]);
    [alert setAttributedTitle:alertTitle];
    [alert setTitle:title];
    [alert setMessage:message];
    [alert setPreferredStyle:UIAlertControllerStyleAlert];
```

pxlcoder commented on Source/WebKit/UIProcess/ios/UIKitUtilities.mm:
```diff
@@ -202,8 +202,15 @@ - (UIViewController *)_wk_viewControllerForFullScreenPresentation
+
+    auto *alert = [[UIAlertController alloc] init];
+    alert.attributedTitle = alertTitle;
+    alert.title = title;
```
Is setting `title` necessary, if we're already setting `attributedTitle`?

youennf replied:
I am not exactly sure but I think we can keep it there since the idea would probably to remove attributedTitle at some point when the underlying bug is fixed.

pxlcoder reviewed (commented):
There are two memory leaks here.

jyavenard commented on Source/WebKit/UIProcess/ios/UIKitUtilities.mm:
```diff
@@ -202,8 +202,15 @@ - (UIViewController *)_wk_viewControllerForFullScreenPresentation
-    auto *alert = [UIAlertController alertControllerWithTitle:title message:message preferredStyle:UIAlertControllerStyleAlert];
-    [alert _setTitleMaximumLineCount:0]; // No limit, we need to make sure the title doesn't get truncated.
+    // FIXME: revisit this once UIKit bug is resolved (see also rdar://101140177).
+    NSMutableAttributedString alertTitle = adoptNS([[NSMutableAttributedString alloc] initWithString:title]);
```
`auto` ?

jyavenard commented on Source/WebKit/UIProcess/ios/UIKitUtilities.mm:
```diff
@@ -202,8 +202,15 @@ - (UIViewController *)_wk_viewControllerForFullScreenPresentation
+    [alertTitle addAttribute:NSFontAttributeName value:[UIFont preferredFontForTextStyle:UIFontTextStyleHeadline] range:NSMakeRange(0, title.length)];
+
+    auto alert = adoptNS([[UIAlertController alloc] init]);
+    alert.get().attributedTitle = alertTitle.get();
```
continuing with the Obj-C style above would have been nicer:
`[alert setAttributedTitle:alertTitle.get()]`
`[alert setBlah:...]`

youennf committed:
createUIAlertController title gets truncated if too long
https://bugs.webkit.org/show_bug.cgi?id=264568
rdar://116345923

Reviewed by Jean-Yves Avenard.

UIKit is truncating titles when they are long, which is probablematic for prompts.
We reuse the same workaround as Safari does for its own prompts.

* Source/WebKit/Platform/spi/ios/UIKitSPI.h:
* Source/WebKit/UIProcess/ios/UIKitUtilities.mm:
(WebKit::createUIAlertController):

Canonical link: https://commits.webkit.org/270760@main
